### PR TITLE
HPCC-27047 K8s Foreign file access

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/templates/eclagent.yaml [hunk 1]
```diff
@@ -67,7 +67,7 @@ data:
 {{- end }}
         spec:
           {{- include "hpcc.placementsByJobTargetType" (dict "root" .root "job" $appJobName "target" .me.name "type" "eclagent") | indent 10 }}
-          serviceAccountName: "hpcc-default"
+          serviceAccountName: "hpcc-agent"
```
Comment (jakesmith) [inline helm/hpcc/templates/eclagent.yaml:70]: should probably be pushed to a separate JIRA/PR.
Hit this when running a test query that forced use of local DFS service from agent.
In that case, the agent needs a service account, to auto discover the local env. dfs service.

File: helm/hpcc/templates/eclagent.yaml [hunk 2]
```diff
@@ -142,7 +142,7 @@ spec:
 {{- end }}
     spec:
       {{- include "hpcc.placementsByPodTargetType" (dict "root" $ "pod" .name "target" .name "type" "eclagent") | indent 6 }}
-      serviceAccountName: {{ .useChildProcesses | default false | ternary "hpcc-default" "hpcc-agent" }}
+      serviceAccountName: "hpcc-agent"
```
Comment (jakesmith) [inline helm/hpcc/templates/eclagent.yaml:145]: as per last comment above

File: helm/hpcc/templates/service-account.yaml [hunk 3]
```diff
@@ -39,10 +39,10 @@ metadata:
   name: hpcc-agent
 rules:
   - apiGroups: [ "" ] # core API group
-    resources: [ "pods" ]
+    resources: [ "pods", "services" ]
     verbs: [ "get", "list" ]
   - apiGroups: [ "batch" ]
-    resources: [ "jobs" ]
+    resources: [ "jobs", "services" ]
```
Comment (jakesmith) [inline helm/hpcc/templates/service-account.yaml:45]: also related to previous commants

**Comment by jakesmith** [comment]
@ghalliday - please give an inital review. I've added some PR comments also

File: fs/dafsserver/dafsserver.cpp [hunk 4]
```diff
@@ -3424,7 +3424,8 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
     Owned<ISocket>      acceptsock;
     Owned<ISocket>      securesock;
     Owned<ISocket>      rowServiceSock;
-
+    Linked<IPropertyTree> componentConfig;
+    bool directIO = true;
```
Comment (ghalliday) [inline fs/dafsserver/dafsserver.cpp:3428]: minor: should this default to false for slightly better security (default to restricted) - although I suspect it is that way for bare-metal defaults?

File: fs/dafsserver/dafsserver.cpp [hunk 5]
```diff
@@ -5017,9 +5018,22 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+            switch (cmd)
+            {
+                case RFCStreamGeneral:
+                case RFCStreamRead:
+                case RFCStreamReadJSON:
+                    break;
+                default:
+                    if (!directIO || client->isRowServiceClient())
+                        throw createDafsException(DAFSERR_cmdstream_unauthorized, "Unauthorized command");
+            }
```
Comment (ghalliday) [inline fs/dafsserver/dafsserver.cpp:5033]: trivial: slightly better to have a break at the end of the default case.  (Some checkers may complain.)

File: fs/dafsserver/dafsserver.cpp [hunk 6]
```diff
@@ -5215,7 +5233,7 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
             if (!securesock)
                 throw createDafsException(DAFSERR_serverinit_failed, "Invalid secure socket");
         }
-
+
```
Comment (ghalliday) [inline fs/dafsserver/dafsserver.cpp:5236]: trivial: trailing space.

File: helm/hpcc/templates/_helpers.tpl [hunk 7]
```diff
@@ -857,8 +860,8 @@ Generate list of available services
 {{- range $.Values.dafilesrv -}}
  {{- if not .disabled -}}
 - name: {{ .name }}
-  type: dafilesrv
-  port: {{ .servicePort | default 7600 }}
+  type: {{ .application | default "stream" }}
+  port: {{ .service.servicePort | default 7600 }}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:864]: Can service be null?  Does it need
$service := .service | default dict
or similar.

**Review by ghalliday — COMMENTED** [review]
@jakesmith looks good.   A few very minor comments

File: fs/dafsserver/dafsserver.cpp (hunk 4 above)
Comment (jakesmith) [reply fs/dafsserver/dafsserver.cpp:3428]: it's default to on for the bare metal case, could reverse it and switch it on if #ifndef _CONTAINERIZED

File: fs/dafsserver/dafsserver.cpp (hunk 5 above)
Comment (jakesmith) [reply fs/dafsserver/dafsserver.cpp:5033]: right, will add.

File: helm/hpcc/templates/_helpers.tpl (hunk 7 above)
Comment (jakesmith) [reply helm/hpcc/templates/_helpers.tpl:864]: service is required, can't be missing.

File: system/jlib/jmisc.cpp [hunk 8]
```diff
@@ -1021,3 +1021,92 @@ jlib_decl char **getSystemEnv()
...
+    {
+        if (input)
+            MLOG(MCinternalError, unknownJob, "Using input %s", input);
+        throw makeStringExceptionV(0, "Failed to run %s: error %u: %s", cmd, ret, error.str());
+    }
+}
+
+std::pair<std::string, unsigned> getExternalService(const char *serviceName)
+{
+    static CTimeLimitedCache<std::string, std::pair<std::string, unsigned>> externalServiceCache;
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/jlib/jmisc.cpp:1048]: will move outside above func.

**Review by ghalliday — APPROVED** [review]
@richardkchapman please review (especially helm changes), and merge if happy.

**Review by richardkchapman — APPROVED** [review]

**Comment by jakesmith** [comment]
@richardkchapman - rebased due to confict after tagging (helmVersion)

**Comment by jakesmith** [comment]
@richardkchapman - All checks have passed after windows build fix - now squashed in.

**Pull request merged**